protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,521 @@
+pragma solidity ^0.8.20;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Temple (templegold/TempleGoldStaking.sol)
+
+
+import { CommonEventsAndErrors } from "contracts/common/CommonEventsAndErrors.sol";
+import { TempleElevatedAccess } from "contracts/v2/access/TempleElevatedAccess.sol";
+import { ITempleGold } from "contracts/interfaces/templegold/ITempleGold.sol";
+import { ITempleGoldStaking } from "contracts/interfaces/templegold/ITempleGoldStaking.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { Pausable } from "@openzeppelin/contracts/utils/Pausable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/** 
+ * @title Temple Gold Staking
+ * @notice Temple Gold Staking contract. Stakers deposit Temple and claim rewards in Temple Gold. 
+ * Temple Gold is distributed to staking contract for stakers on mint.
+ * Duration for distributing staking rewards is set with `setRewardDuration`. An unstake period is used
+ * to encourage longer staking times.
+ */
+contract TempleGoldStaking is ITempleGoldStaking, TempleElevatedAccess, Pausable {
+    using SafeERC20 for IERC20;
+
+    /// @notice The staking token. Temple
+    IERC20 public immutable override stakingToken;
+    /// @notice Reward token. Temple Gold
+    IERC20 public immutable override rewardToken;
+
+    /// @notice Distribution starter
+    address public override distributionStarter;
+    /// @notice Week length
+    uint256 constant public WEEK_LENGTH = 7 days;
+    /// @notice Total supply of staking token
+    uint256 public override totalSupply;
+
+    /// @notice Time tracking
+    uint256 public override periodFinish;
+    /// @notice Last rewards update time
+    uint256 public override lastUpdateTime;
+
+    /// @notice Store next reward amount for next epoch
+    uint256 public override nextRewardAmount;
+    /// @notice Duration for rewards distribution
+    uint256 public override rewardDuration;
+    /// @notice Cooldown time before next distribution of rewards
+    /// @dev If set to zero, rewards distribution is callable any time 
+    uint160 public override rewardDistributionCoolDown;
+    /// @notice Timestamp for last reward notification
+    uint96 public override lastRewardNotificationTimestamp;
+    /// @notice Cooldown time before unstake
+    uint32 public override unstakeCooldown;
```

Comment:
`unstakeCooldown` is not initialized until calling `setUnstakeCooldown()`, which means that the default is "no unstake cooldown" until it is set. 

But I'm sure you're aware of it. Not a big deal anyways